internal/blockchain/cardano/cardano.go

```diff
@@ -0,0 +1,653 @@
+	}
+
+	return statusResponse, nil
+}
+
+func (c *Cardano) afterConnect(ctx context.Context, w wsclient.WSClient) error {
+	// Send a subscribe to our topic after each connect/reconnect
+	b, _ := json.Marshal(&cardanoWSCommandPayload{
+		Type:  "listen",
+		Topic: c.pluginTopic,
+	})
+	err := w.Send(ctx, b)
+	if err == nil {
+		b, _ = json.Marshal(&cardanoWSCommandPayload{
+			Type: "listenreplies",
+		})
+		err = w.Send(ctx, b)
+	}
+	return err
+}
+
+func (c *Cardano) recoverFFI(ctx context.Context, parsedMethod interface{}) (*fftypes.FFIMethod, []*fftypes.FFIError, error) {
+	methodInfo, ok := parsedMethod.(*ffiMethodAndErrors)
+	if !ok || methodInfo.method == nil || methodInfo.method.Name == "" {
+		return nil, nil, i18n.NewError(ctx, coremsgs.MsgUnexpectedInterfaceType, parsedMethod)
+	}
+	return methodInfo.method, methodInfo.errors, nil
+}
+
+func (c *Cardano) eventLoop() {
+	defer c.wsconn.Close()
+	defer close(c.closed)
+	l := log.L(c.ctx).WithField("role", "event-loop")
+	ctx := log.WithLogger(c.ctx, l)
+	for {
+		select {
+		case <-ctx.Done():
+			l.Debugf("Event loop exiting (context cancelled)")
+			return
+		case msgBytes, ok := <-c.wsconn.Receive():
+			if !ok {
+				l.Debugf("Event loop exiting (receive channel closed). Terminating server!")
+				c.cancelCtx()
+				return
+			}
+
+			var msgParsed interface{}
+			err := json.Unmarshal(msgBytes, &msgParsed)
+			if err != nil {
+				l.Errorf("Message cannot be parsed as JSON: %s\n%s", err, string(msgBytes))
+				continue // Swallow this and move on
+			}
+			switch msgTyped := msgParsed.(type) {
+			case []interface{}:
+				err = c.handleMessageBatch(ctx, 0, msgTyped)
+				if err == nil {
+					ack, _ := json.Marshal(&cardanoWSCommandPayload{
+						Type:  "ack",
+						Topic: c.pluginTopic,
+					})
+					err = c.wsconn.Send(ctx, ack)
+				}
+			case map[string]interface{}:
+				isBatch := false
+				if batchNumber, ok := msgTyped["batchNumber"].(float64); ok {
+					if events, ok := msgTyped["events"].([]interface{}); ok {
+						// FFTM delivery with a batch number to use in the ack
+						isBatch = true
+						err = c.handleMessageBatch(ctx, (int64)(batchNumber), events)
+						// Errors processing messages are converted into nacks
+						ackOrNack := &cardanoWSCommandPayload{
+							Topic:       c.pluginTopic,
+							BatchNumber: int64(batchNumber),
+						}
+						if err == nil {
+							ackOrNack.Type = "ack"
+						} else {
+							log.L(ctx).Errorf("Rejecting batch due error: %s", err)
+							ackOrNack.Type = "error"
+							ackOrNack.Message = err.Error()
+						}
+						b, _ := json.Marshal(&ackOrNack)
+						err = c.wsconn.Send(ctx, b)
+					}
+				}
+				if !isBatch {
+					var receipt common.BlockchainReceiptNotification
+					_ = json.Unmarshal(msgBytes, &receipt)
+
+					err := common.HandleReceipt(ctx, "", c, &receipt, c.callbacks)
+					if err != nil {
+						l.Errorf("Failed to process receipt: %+v", msgTyped)
+					}
+				}
```

Comment:
I see that this is heavily inspired from FFTM and the evm plugin - I would recommend in the future moving to a durable event stream here where the receipts have to be ack'd back as part of a batch.

Reply:
Agreed - probably as a follow up. I'm building a framework at moment to allow for resilient ack-ing and get confirmation from FireFly core it has processed things correctly